Validate parameters in config file

perldork opened the pull request:
The command line parser validates that parameters that require integer options are integers, the configuration file validator was not doing this same type of validation.

* Ensure verbose is either true or false
* Ensure integer parameters are valid integers

davydovanton commented on lib/sidekiq/cli.rb:
```diff
@@ -269,6 +269,19 @@ def validate!
+      errors = []
+      [:concurrency, :index, :timeout].each do |arg|
+        if options.has_key? arg
+          Integer(options[arg]) rescue errors.push("#{arg.to_s} must be an integer")
```
Please, delete unnecessary `to_s` method. Ruby use `to_s` by default for interpolation. You can check this in irb session:
```
irb(main):001:0> "#{[1,2,3]}"
=> "[1, 2, 3]"
irb(main):002:0> "#{[1,2,3].to_s}"
=> "[1, 2, 3]"
irb(main):003:0> "#{{a: 1}}"
=> "{:a=>1}"
irb(main):004:0> "#{{a: 1}.to_s}"
=> "{:a=>1}"
```

davydovanton replied:
And also I prefer
```ruby
errors.push unless options[arg].is_a? Integer
```
instead
``` ruby
Integer(options[arg]) rescue ...
```

@mperham what do you think?

mperham replied:
I prefer not to use exceptions at all for validation.

davydovanton commented on lib/sidekiq/cli.rb:
```diff
@@ -269,6 +269,19 @@ def validate!
+      end
+      if options.has_key? :verbose
+        val = options[:verbose]
+        errors.push("verbose must be true or false") unless val == true || val == false
```
Also you can use `!!val == val` code for check bool type:
```
irb(main):014:0> foo = 1
=> 1
irb(main):015:0> !!foo == foo
=> false
irb(main):016:0> foo = 'test'
=> "test"
irb(main):017:0> !!foo == foo
=> false
irb(main):018:0> foo = []
=> []
irb(main):019:0> !!foo == foo
=> false
irb(main):020:0> foo = false
=> false
irb(main):021:0> !!foo == foo
=> true
```

mperham replied:
verbose is a flag given to us by optparse.  There's no need to validate it at all.

mperham commented on lib/sidekiq/cli.rb:
```diff
@@ -269,6 +269,19 @@ def validate!
         die(1)
       end
+
+      errors = []
```
No need to collect all errors, just fail fast on the first error.

perldork replied:
Isn't it a nicer user experience to see all errors at once? For me as a user I prefer that instead of fix, fail, fix, fail ..

mperham replied:
As a developer, if you see one error for your config values, you can pretty safely extrapolate that similar values will also be invalid.  I prefer the simpler impl in this case.

syndbg replied:
In favor of collecting errors, wouldn't a setting as :fail_on_first (default - true) solve this problem and make both sides happy?

perldork replied:
It's only 2 parameters with 2 possible error conditions - fail fast works
for me. Patch updated.

On Fri, May 15, 2015 at 5:00 AM Anton Antonov <notifications@github.com>
wrote:

> In lib/sidekiq/cli.rb
> <https://github.com/mperham/sidekiq/pull/2352#discussion_r30394222>:
>
> > @@ -269,6 +269,19 @@ def validate!
> >          logger.info @parser
> >          die(1)
> >        end
> > +
> > +      errors = []
>
> In favor of collecting errors, wouldn't a setting as :fail_on_first
> (default - true) solve this problem and make both sides happy?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mperham/sidekiq/pull/2352/files#r30394222>.
>

mperham commented on lib/sidekiq/cli.rb:
```diff
@@ -269,6 +269,19 @@ def validate!
       end
+
+      errors = []
+      [:concurrency, :index, :timeout].each do |arg|
```
`index` is not a field you would pass in the config file (since it's process-specific) so you don't need to validate it.  0 is not a legal value for concurrency and timeout so you can do this:

```ruby
raise ArgumentError, "Invalid timeout value #{options[:timeout]} in #{options[:config_file]}" if options.has_key?(:timeout) && options[:timeout].to_i == 0
raise ArgumentError, "Invalid concurrency value #{options[:concurrency]} in #{options[:config_file]}" if options.has_key?(:concurrency) && options[:concurrency].to_i == 0
```

perldork commented:
Thanks for all the feedback!
* Removed exception based validation
* Cleaned up error messages to be more clear to user
* Added range validation for timeout and concurrency
* Removed :index

perldork commented:
Removed the :verbose flag processing as well per Mike's comments.

davydovanton commented:
:cool:
